[CALCITE-2822] Allow MultiJoin rules with any project/filter

siddharthteotia opened the pull request:
Allow MultiJoin rules to be used with any subclass of Project or Filter
instead of only assuming LogicalProject and LogicalFilter

siddharthteotia commented:
@vlsi , I have addressed the review comments -- added unit tests and new constructor to not break the backward compatibility. Please review when you get a chance.

siddharthteotia commented:
Addressed build issues

vlsi commented on core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java:
```diff
@@ -4292,6 +4293,21 @@ private Sql checkSubQuery(String sql) {
+    final FilterMultiJoinMergeRule filterMultiJoinMergeRule =
+        new FilterMultiJoinMergeRule(LogicalFilter.class, RelFactories.LOGICAL_BUILDER);
+    final ProjectMultiJoinMergeRule projectMultiJoinMergeRule =
+        new ProjectMultiJoinMergeRule(LogicalProject.class, RelFactories.LOGICAL_BUILDER);
```
Well, LogicalProject was the default class, so this test does not seem to validate anything new. Does it just duplicate the existing tests?

siddharthteotia replied:
I added a new constructor in FilterMultiJoinMergeRule and ProjectMultiJoinMergeRule that take a generic Filter.class and Project.class as arguments respectively. The unit test just uses the new constructors with LogicalFilter and LogicalProject

vlsi replied:
Assume the implementation is "ignore the parameter and always use `LogicalProject.class`"
Does the test catch that?

siddharthteotia commented:
@vlsi , I have rebased and added a test case to test Filter and Project MultiJoin merge rules work with any implementations of Filter and Project as opposed to just LogicalFilter and LogicalProject. For this I added CustomFilter and CustomProject as implementations of Filter and Project interfaces. Similarly, I had to pass in a custom RelBuilderFactory to SqltoRelConverter config as by default it uses RelFactories.LOGICAL_BUILDER which has all the default factories. The test case has implementations of Custom filter and project factory.

One problem I have been facing is that SqlToRelConverter has hardcoded the filter factory to be used as default filter factor in RelFactories and thus the custom implementation doesn't get used for Filter. If you look at the test case plan in RelOptRulesTest.xml, you can see that the plan now has "MyProject" instead of "LogicalProject" but it is still having "LogicalFilter" because of this hardcoded factory in SqlToRelConverter at line#990

I couldn't really find a way of extracting the configured filter factor in SqlToRelConverter config -- something like config.getRelBuilderFactory.getFilterFactory(). AFAIK, there is probably no way of getting the specific factory class and using that instead of hardcoding to LogicalFilter factory.

Can you please advise how this can be fixed?

zabetak commented:
SqlToRelConverter should use the RelBuilderFactory provided in the configuration. As you observed there are places where this does not happen. To me this appears like a bug so you could maybe fix it as part of this PR. With a first glance it does not look so complicated to replace hardcoded factories with those in the configuration.

zabetak commented on core/src/main/java/org/apache/calcite/plan/RelOptUtil.java:
```diff
@@ -2611,12 +2611,12 @@ private static RexShuttle pushShuttle(final Project project) {
    * {@link org.apache.calcite.rel.rules.MultiJoin}.
```
The previous line mentions LogicalProject which is no longer true.

siddharthteotia replied:
done

zabetak commented on core/src/main/java/org/apache/calcite/rel/rules/FilterMultiJoinMergeRule.java:
```diff
@@ -52,10 +55,24 @@ public FilterMultiJoinMergeRule(RelBuilderFactory relBuilderFactory) {
+   * @param filterClass filter class
+   * @param relBuilderFactory builder factory for relational expressions
+   */
+  public FilterMultiJoinMergeRule(Class<? extends Filter> filterClass,
```
The main Javadoc of the class (around line 34) has to be adapted too.

siddharthteotia replied:
done

zabetak commented on core/src/main/java/org/apache/calcite/rel/rules/FilterMultiJoinMergeRule.java:
```diff
@@ -43,7 +44,9 @@
+   * @param relBuilderFactory builder factory for relational expressions
    */
   public FilterMultiJoinMergeRule(RelBuilderFactory relBuilderFactory) {
     super(
```
I think it is better to use the telescoping constructor pattern here.

siddharthteotia replied:
done

siddharthteotia commented:
> SqlToRelConverter should use the RelBuilderFactory provided in the configuration. As you observed there are places where this does not happen. To me this appears like a bug so you could maybe fix it as part of this PR. With a first glance it does not look so complicated to replace hardcoded factories with those in the configuration.

@zabetak , One thing that I wasn't really able to see is how to get a particular factory out of RelBuilderFactory. As you said, SqlToRelConverter's config has RelBuilderFactory available as config.getRelBuilderFactory()... my question is how do I now get the specific factory out of this? Like, for the particular hard-coded case I pointed to... I need something like config.getRelBuilderFactory().getFilterFactory()

There is no such method or equivalent to getFilterFactory().

zabetak commented on core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java:
```diff
@@ -4321,12 +4331,12 @@ private Sql checkSubQuery(String sql) {
         .scan("EMP")
         .filter(
-            builder.call(SqlStdOperatorTable.EQUALS,
-                builder.field("EMPNO"), builder.literal(10)))
```
Did you modify the formatting on purpose?

siddharthteotia replied:
Sorry this change was not on purpose

siddharthteotia replied:
done

zabetak commented:
> I need something like config.getRelBuilderFactory().getFilterFactory()

The factories are inside the created RelBuilder. Can't you use something like:
`final RelNode filter = relBuilder.filter(convertedWhere2);`
or
 `relBuilder.push(bb.root);`
 `final RelNode filter = relBuilder.filter(convertedWhere2);`
for instance?

siddharthteotia commented:
@zabetak , I have addressed the comments

> > I need something like config.getRelBuilderFactory().getFilterFactory()
> 
> The factories are inside the created RelBuilder. Can't you use something like:
> `final RelNode filter = relBuilder.filter(convertedWhere2);`
> or
> `relBuilder.push(bb.root);`
> `final RelNode filter = relBuilder.filter(convertedWhere2);`
> for instance?
> I didn't really look how the relBuilder is used previously but I guess should be something similar.

Thanks @zabetak , I was able to use the factory in RelBuilder. I have addressed the other comments as well. Please review when you get a chance

siddharthteotia commented on core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableRel.java:
```diff
@@ -27,9 +26,6 @@
  */
 public interface EnumerableRel
     extends RelNode {
-  RelFactories.FilterFactory FILTER_FACTORY = EnumerableFilter::create;
```
done

siddharthteotia commented on core/src/main/java/org/apache/calcite/tools/RelBuilder.java:
```diff
@@ -2623,6 +2623,10 @@ public RexNode visitInputRef(RexInputRef inputRef) {
     }
   }
+
+  public RelFactories.FilterFactory getFilterFactory() {
```
removed

siddharthteotia commented on core/src/main/java/org/apache/calcite/rel/core/RelFactories.java:
```diff
@@ -295,6 +296,10 @@ public RelNode createAggregate(RelNode input, boolean indicator,
     RelNode createFilter(RelNode input, RexNode condition);
+
+    /** Creates a filter */
+    RelNode createFilter(RelNode input, RexNode condition,
```
removed

siddharthteotia commented:
> Thanks for the changes @siddharthteotia ! However, I noticed that there are a few changes on public API that are useful only for the new tests and some of them also break backward compatibility. As far as I can see for the tests, the only thing that you need is a plan with a custom filter/project, join, plus some table scans. I don't think it is necessary to pass from the SqlToRelConverter to do this. (It would be nice but we should not stress too much). You can use a custom RelBuilder to create the plan directly and apply the rules on it. Like that you will not have to introduce changes in so many places. Let me know how does this look.

@zabetak , I have addressed your comments by writing tests in a different manner. Please take a look when you get a chance. Also rebased the patch.

Cc @laurentgo

siddharthteotia commented:
Build is passing. @zabetak , @laurentgo , please take a look when you have some time.

rubenada commented on core/src/main/java/org/apache/calcite/rel/rules/MultiJoinProjectTransposeRule.java:
```diff
@@ -128,7 +129,7 @@ protected LogicalProject getRightChild(RelOptRuleCall call) {
   // override JoinProjectTransposeRule
```
Minor detail: maybe this is a good opportunity to replace all four comments `// override JoinProjectTransposeRule` in this class with the proper annotation `@Override`

laurentgo replied:
I'd happy to post a patch to fix it: I think @siddharthteotia has been really patient, so let's not abuse his time

rubenada replied:
Sure, no problem.

laurentgo commented:
@zabetak I think your comments have been addressed. FYI I plan to merge @siddharthteotia change today.

zabetak commented:
I will have a look right now, thanks for the reminder!
